desc/compute/utils.py

```diff
@@ -1204,6 +1204,112 @@ def surface_integrals_transform(grid, surface_label="rho"):
     return surface_integrals_map(grid, surface_label, expand_out=False)
 
 
+def surface_variance(
+    grid,
+    q,
+    weights=jnp.array([1.0]),
+    unbiased=True,
+    surface_label="rho",
+    expand_out=True,
+):
+    r"""Compute the weighted sample variance of ``q`` on each surface of the grid.
+
+    Computes the following quantity on each surface of the grid.
+
+    .. math::
+        \frac{n_e}{n_e - b}
+        \frac{ \sum_{i=1}^{n} (q_i - \bar{q})^2 w_i }{ \sum_{i=1}^{n} w_i }
+
+    where
+    :math:`w_i` is the weight assigned to :math:`q_i` given by the product
+    of ``weights[i]`` and the differential surface area element (not already
+    weighted by the area Jacobian) at the node where ``q[i]`` is evaluated,
+    :math:`\bar{q}` is the weighted mean of :math:`q`,
+    :math:`b` is 0 if the biased sample variance is to be returned and 1 otherwise,
+    :math:`n` is the number of samples on a surface, and
+    :math:`n_e` is the effective number of samples on a surface defined as
+
+    .. math::
+        (\sum_{i=1}^{n} w_i)^2 / (\sum_{i=1}^{n} w_i^2)
+
+    As the weights :math:`w_i` approach each other, :math:`n_e` approaches
+    :math:`n`, and the output converges to
+
+    .. math::
+        \frac{1}{n - b} \sum_{i=1}^{n} (q_i - \bar{q})^2
+
+    Notes
+    -----
+        There are three different methods to unbias the variance of a weighted
+        sample so that the computed variance better estimates the true variance.
+        Whether the method is correct for a particular use case depends on what
+        the weights assigned to each sample represent.
+
+        This function implements the first case, where the weights are not random
+        and are intended to assign more weight to some samples for reasons
+        unrelated to differences in uncertainty between samples. See
+        https://en.wikipedia.org/wiki/Weighted_arithmetic_mean#Reliability_weights.
+
+        The second case is when the weights are intended to assign more weight
+        to samples with less uncertainty. See
+        https://en.wikipedia.org/wiki/Inverse-variance_weighting.
+        The unbiased sample variance for this case is obtained by replacing the
+        effective number of samples in the formula this function implements,
+        :math:`n_e`, with the actual number of samples :math:`n`.
+
+        The third case is when the weights denote the integer frequency of each
+        sample. See
+        https://en.wikipedia.org/wiki/Weighted_arithmetic_mean#Frequency_weights.
```

Comment:
FYI, If you read through this link I think you'll find that wikipedia has some errors. That formula should only be used on integer frequeny weights. They suggest that you can still use the formula when the frequency weights are normalized, but in addition to loosing ability to unbias, normalizing to 1 blows up their expression as the denominator is 0.